### DISCUSSION TITLE
How can I use core 1 to assign work to core 0

### DISCUSSION THREAD

**Discussion by GM-Script-Writer-62850** [micropython/micropython#10434] [General]
Note that CPU cores are `[0,1]` running on a [W5500_EVB_PICO](https://micropython.org/download/W5500_EVB_PICO/)

I would like to have core 1 doing time sensitive polling work using `_thread`

When core 1 finds work I want to send it to core 0 where core 0 is using uasyncio, i know `_thread` is buggy so i want to keep it simple core running on core 1 and let core 0 do the fancy stuff

I was trying to use interrupts on core 0 and i am having a lot of issues with switch bounce and EMI, so i am just going to use polling ever ms, dealing with switch bounce is one thing but EMI...

**Comment by GM-Script-Writer-62850** [comment]
Surprised I do not have a reply yet, so i have this dumb hack solution, I really hope I do not have a EMI issue with this case that will SUCK...

Sadly this is still a interrupt so it will stop other work from being done, should only be a issue if core 0 is reading a DHT22 sensor... but since the results are not time sensitive that is not a issue if i have to re-poll the sensor once and a while

but i can store a assignment in a variable toggle the pin triggering the interrupt then core0 will do the work i want when i want it done
```
from machine import Pin
from _thread import start_new_thread as newThread
from time import sleep

p=Pin(0,Pin.OUT,value=0)# A spare/NC pin

def work(pin):
	print('hi')

p.irq(handler=work, trigger=Pin.IRQ_RISING | Pin.IRQ_FALLING)

def core1():
	while True:
		sleep(1)
		p.toggle()

newThread(core1,())

while True:
	print("waiting")
	sleep(1)

```

**Reply by cve2022** [reply]
> I would like to have core 1 doing time sensitive polling work using _thread


Hi, I have a similar(ish) question, take a look at #10371 . 

I still have some tests to do related to the aforementioned discussion, but according to my understanding, I will have to leave the core 0 for the realtime stuff and core 1 will do other not time critical tasks. I think it should work for your use case as well and you will need to use some IPC resource for the task in core 0 to send whatever it collected to the task/process/coroutine running on core 1 for post processing.

HTH.

**Reply by GM-Script-Writer-62850** [reply]
hold up a second the PICO has SMT? i though it was 2 cores 2 threads, not 2 cores 4 threads

**Reply by cve2022** [reply]
> hold up a second the PICO has SMT? i though it was 2 cores 2 threads, not 2 cores 4 threads

Nope. With _thread you have only two threads, one thread per core.

You may find [uasyncio ](https://docs.micropython.org/en/latest/library/uasyncio.html)of interest. It simulates multithreading by the use of coroutines. There is a thorough tutorial on it: [https://github.com/peterhinch/micropython-async](https://github.com/peterhinch/micropython-async).

In the case of RP2, I believe you can combine _thread and uasyncio to have the benefit of using the two available cores of the RP2 and the concurrency capabilities of uasyncio to code a fairly complex application. Depending on your systems requirements, you may find asyncio more than sufficient though.

HTH.

**Reply by GM-Script-Writer-62850** [reply]
i am somewhat familiar with uasyncio and do intended to use it, i just also make use of thread, i know how to make use of sleep and await and i think i know how to use yield

i have not tired to use it on core 1, only used it on core 0 so far, hopefully it is not bugged on _thread and i can use it on both cores

**Comment by jimmo** [comment]
To answer the general question -- if you want to make the second core trigger asyncio tasks in the first core, then asyncio.ThreadSafeFlag is what you need.

> i know _thread is buggy so i want to keep it simple core running on core 1 and let core 0 do the fancy stuff

@GM-Script-Writer-62850 can you provide a reference or background for this? I feel like I keep hearing this repeated but haven't been able to get anyone to demonstrate a repro for this (since we fixed stuff several months ago).

> > hold up a second the PICO has SMT? i though it was 2 cores 2 threads, not 2 cores 4 threads
> 
> Nope. With _thread you have only two threads, one thread per core.

Just to clarify because despite the "Nope." from @cve2022 it seems like you're agreeing. On the rp2040 there is two cores, and one thread per core (i.e. you may optionally create one additional thread).

**Reply by GM-Script-Writer-62850** [reply]
> can you provide a reference or background for this? I feel like I keep hearing this repeated but haven't been able to get anyone to demonstrate a repro for this (since we fixed stuff several months ago).

https://docs.micropython.org/en/latest/library/_thread.html (maybe this is in dire need of updating)

issues i have run into:

* interrupts not acting as expected when set on in `_thread`
* not sure if this is `_thread`'s fault but... bug #10448

do you have some example code using `ThreadSafeFlag`?

**Reply by peterhinch** [reply]
See [this doc](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/THREADING.md) for information on using `uasyncio` with threaded code. Also [this tutorial section](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/TUTORIAL.md#36-threadsafeflag) on `ThreadSafeFlag`. It includes code samples.

**Reply by GM-Script-Writer-62850** [reply]
an async interrupt... does this work inside thread?

i am assuming i can do this: `_thread.start_new_thread(uasyncio.run,(main()))`

i ran into issues running stuff on core 0 while using _thread, so `_thread` gets the complex code

* https://github.com/micropython/micropython/issues/10482
* https://github.com/micropython/micropython/issues/10448

**Reply by peterhinch** [reply]
Currently `uasyncio` only runs on core 0. @dpgeorge has mentioned the possibility of running a copy on each core, but as far as I know this is not yet supported. Also interrupt service routines must be declared on core 0.

**Reply by GM-Script-Writer-62850** [reply]
i think i have my uasyncio running on core 1 and my main non-uasyncio loop on core 0, i assume this would not appear to work otherwise

```
import time
import _thread
import uasyncio

async def main():
  while 1:
    print('core1')
    await uasyncio.sleep(1)

def _main():
  uasyncio.run(main())
_thread.start_new_thread(_main,())

while 1:
  print('core0')
  time.sleep(1)
```

**Reply by cve2022** [reply]
> Currently `uasyncio` only runs on core 0. @dpgeorge has mentioned the possibility of running a copy on each core, but as far as I know this is not yet supported. Also interrupt service routines must be declared on core 0.

Is asyncio hardwired to core 0 or you state that because timers and interrupts run on core 0? If the latter, could we have asyncio  running on core 1 if we do not use timing functions but instead use e.g. events from core 0 to waking up coroutines on core 1 (if that is possible?)

**Reply by jimmo** [reply]
> Is asyncio hardwired to core 0 

@peterhinch @GM-Script-Writer-62850 @cve2022 asyncio is completely oblivious to threads. You can run it on either core, but you cannot run it on both cores concurrently because it has global state.

**Reply by GM-Script-Writer-62850** [reply]
but can you use it on both if you do this?
```
import uasyncio # for core 0
import uasyncio as _uasyncio  # for core 1
```

**Reply by jimmo** [reply]
No, that is not how importing modules works.

**Reply by GM-Script-Writer-62850** [reply]
Wait a min if I can only use uasync on one core and have it on core 1, if i use `uasyncio.createTask` on core 0 will it run on core 1

**Reply by peterhinch** [reply]
No. `uasyncio.createTask` must run on the same core as `uasyncio`.

**Reply by jimmo** [reply]
To clarify this -- `create_task` will mutate asyncio's internal state. As above, asyncio is completely thread-oblivious, so this mutation can race other internal operations (such as scheduling).

If you have asyncio running on one core, and you want to trigger something from a different core (or context, such as IRQ) you _must_ use ThreadSafeFlag (or something built on that).
(In more detail: you could of course just set a global variable and poll that from your asyncio code, but this is less efficient than using ThreadSafeFlag)

FWIW, Python has https://docs.python.org/3/library/asyncio-task.html#asyncio.run_coroutine_threadsafe which is something we might consider adding.

**Reply by peterhinch** [reply]
@GM-Script-Writer-62850 To implement what you want you need something like this:
```python
import uasyncio as asyncio
tsf = asyncio.ThreadSafeFlag()  # Shared global
async def run_on_demand():
    while True:
        await tsf.wait()
        # Run payload code
```
Code in the other thread issues `tsf.set()` when it wants the payload to run.

**Reply by GM-Script-Writer-62850** [reply]
Thanks, that was something i wanted to know how to use, do i need to use a `tsf.unset()` or is `tfs.set()` consumed by `tsf.wait()`
